src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, editor.GetChangedRoot());
+
+            // Adds an argument which is an invocation of the newly created method to the callsites
+            // of the method invocations where a parameter was added.
+            // Example:
+            // public void M(int x, int y)
+            // {
+            //     int f = x * y; // highlight this expression
+            // }
+            // 
+            // public void M1()
+            // {
+            //     M(5, 6);
+            // }
+            //
+            // ---------------------------------------------------->
+            // 
+            // public int M_f(int x, int y)
+            // {
+            //     return x * y;
+            // }
+            // 
+            // public void M(int x, int y)
+            // {
+            //     int f = x * y;
+            // }
+            //
+            // public void M1()
+            // {
+            //     M(5, 6, M_f(5, 6)); // This is the generated invocation which is a new argument at the call site
+            // }
+            static TInvocationExpressionSyntax GenerateNewInvocationExpressionForTrampoline(ISyntaxFactsService syntaxFacts,
+                SyntaxEditor editor, SyntaxNode currentInvocation, TInvocationExpressionSyntax invocationExpression,
+                string newMethodIdentifier, int insertionIndex)
+            {
+                var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(currentInvocation);
+                var methodName = editor.Generator.IdentifierName(newMethodIdentifier);
+                var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+                var newMethodInvocation = editor.Generator.InvocationExpression(methodName, invocationArguments);
+                var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+                return (TInvocationExpressionSyntax)editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+            }
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private static async Task<Solution> RewriteSolutionWithNewMethodAsync(
+            Document originalDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            CancellationToken cancellationToken)
+        {
```

Comment:
nit.  it woudl probably be good to move this abouve the helper methods it calls.  i found the review confusing as i hit the helpers first, and didn't really know how they wer ecalled.  while not always possible consider the top->to->bottom flow of code, and have it be similar to the *Execution* flow.  i.e. a method that is only called by another method should appear lower than it to make it easier to read and understand the code.

Reply:
good point, moved upwards